src/WebApplication.php

```diff
@@ -97,115 +111,122 @@ public function doExecute()
+			'optionName'        => $resourceName,
+			'viewName'          => $input->getString('view'),
+			'webserviceVersion' => $version,
+			'webserviceClient'  => $client,
+			'method'            => strtoupper($input->getCmd('method', $input->getMethod())),
+			'task'              => $input->getCmd('task'),
+			'data'              => $this->getPostedData(),
+			'dataGet'           => $input->getArray(),
+			'accessToken'       => $input->getString($this->get('webservices.oauth2_token_param_name', 'access_token')),
+			'id'                => $input->getString('id'),
+			'absoluteHrefs'     => $input->getBool('absoluteHrefs', true),
+		);
+
+		$rendererOptions = [
+			'charset'		=> 'utf-8',
+			'language'		=> 'en-GB',
+			'direction'		=> 'ltr',
+			'link'			=> '',
+			'base'			=> '',
+			'absoluteHrefs'	=> $input->getBool('absoluteHrefs', true),
+			'uriParams'		=> [],
+			'resourceName'	=> $resourceName,
+		];
 
 		try
 		{
-			$this->clearHeaders();
-			$webserviceClient = $input->getString('webserviceClient', 'administrator');
-			$optionName       = $input->getString('option');
-			$optionName       = strpos($optionName, 'com_') === 0 ? substr($optionName, 4) : $optionName;
-			$viewName         = $input->getString('view');
-			$version          = $input->getString('webserviceVersion');
-
-			$token = $input->getString($this->get('webservices.oauth2_token_param_name', 'access_token'));
-			$apiName = ucfirst($apiName);
-			$method  = strtoupper($input->getMethod());
-			$task    = $this->getTask();
-			$data    = $this->getPostedData();
-			$dataGet = $input->getArray();
-
-			$options = array(
-				'api'               => $apiName,
-				'optionName'        => $optionName,
-				'viewName'          => $viewName,
-				'webserviceVersion' => $version,
-				'webserviceClient'  => $webserviceClient,
-				'method'            => $method,
-				'task'              => $task,
-				'data'              => $data,
-				'dataGet'           => $dataGet,
-				'accessToken'       => $token,
-				'format'            => $input->getString('format', $this->get('webservices.webservices_default_format', 'hal')),
-				'id'                => $input->getString('id'),
-				'absoluteHrefs'     => $input->getBool('absoluteHrefs', true),
-			);
-
-			$apiClass = 'Joomla\\Webservices\\Api\\' . $apiName . '\\' . $apiName;
-
-			if (!class_exists($apiClass))
-			{
-				throw new \RuntimeException($text->sprintf('LIB_WEBSERVICES_API_UNABLE_TO_LOAD_API', $options['api']));
-			}
-
-			try
-			{
-				/** @var \Joomla\Webservices\Api\ApiBase $api */
-				$api = new $apiClass($this->getContainer(), new Registry($options));
-			}
-			catch (\RuntimeException $e)
-			{
-				throw new \RuntimeException($text->sprintf('LIB_WEBSERVICES_API_UNABLE_TO_CONNECT_TO_API', $e->getMessage()));
-			}
-
-			// Run the api task
-			$api->execute();
-
-			// Display output
-			$api->render();
+			// Instantiate a renderer, based on content negotiation, then defer to the API class.
+			$this->container
```

Comment:
Why are we setting the API Provider here and then only using 2 lines later!? If it's just to reduce code duplication between the cli and web application class use a trait (we have so many other 5.4 dependencies at this point including your content negotiator library...) otherwise what is the thinking?